src/Vehicle/Vehicle.cc

```diff
@@ -2104,40 +2100,26 @@ void Vehicle::_linkInactiveOrDeleted(LinkInterface* link)
     }
 }
 
-bool Vehicle::sendMessageOnLink(LinkInterface* link, mavlink_message_t message)
+bool Vehicle::sendMessageOnLinkThreadSafe(LinkInterface* link, mavlink_message_t message)
 {
-    if (!link || !_links.contains(link) || !link->isConnected()) {
-        return false;
-    }
-
-    emit _sendMessageOnLinkOnThread(link, message);
-
-    return true;
-}
+    QMutexLocker lock(&_sendMessageOnLinkMutex);
 
```

Comment:
Is this lock needed given that `link->writeBytesThreadSafe()` has its own lock? What is it protecting?

Reply:
Makes sure this stuff is thread safe as well:
```
    if (!link->isConnected()) {
        return false;
    }

    // Give the plugin a chance to adjust
    _firmwarePlugin->adjustOutgoingMavlinkMessage(this, link, &message);
```

Reply:
Since the joystick send will come through this path from another thread.

Reply:
This `link->isConnected()` probably isn't too problematic. But this `_firmwarePlugin->adjustOutgoingMavlinkMessage` is since it maintains a mavlink channel status if I remember correctly.

Reply:
Yup, It doesn't look like it is needed. The `adjustOutgoingMavlinkMessage()` function is 99.9% void. The only time something actually is done is for [this](https://github.com/mavlink/qgroundcontrol/blob/master/src/FirmwarePlugin/APM/APMFirmwarePlugin.cc#L511-L518):
```
    case MAVLINK_MSG_ID_PARAM_SET:
        _handleOutgoingParamSet(vehicle, outgoingLink, message);
```
Might as well put a mutex in there rather than annoying the other 99.9% of the time this never used.

Reply:
The problem is that it's a firmwareplugin override so you have no idea what can happen in there in a custom build. Without a change to the FirmwarePlugin method to mark as threadsafe required and likely a name change to prevent old unthreadsafe code from leaking through it needed. I can certainly do that. Seems like a better way to go intstead of the upper layers doing things which won't be needed most of the time.